src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+
+        Node indexNode = getNode(index);
+        T returnNode = (T)indexNode.value;
+        indexNode.value = value;
+
+        return returnNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+
+        Node returnNode = getNode(index);
+        unlink(returnNode);
+        size--;
+
+        return (T)returnNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value != null && currentNode.value.equals(object)
+                    || currentNode.value == null && object == null) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
+
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private class Node<T> {
+        private Node prev;
+        private Node next;
+        private T value;
+
+        public Node(Node prev, T value, Node next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void unlink(Node node) {
+        if (node.next == null && node.prev == null) {
+            head = null;
+            tail = null;
+            return;
+        }
+
+        if (node.next == null) {
+            node.prev.next = null;
+            tail = node.prev;
+            return;
+        }
+
+        if (node.prev == null) {
+            node.next.prev = null;
+            head = node.next;
+            return;
+        }
+
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("index "
+                    + index + " is equals or out of size " + size);
+        }
+    }
+
+    private void addHead(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = new Node<>(null, value, null);
+            size++;
+            return;
+        }
+        Node node = new Node<>(head, value, null);
+        tail = node;
+        tail.prev = head;
+        head.next = node;
+        size++;
+    }
+
+    private void addIfIndex0(T value, int index) {
+        Node node = new Node<>(null, value, head);
+        head.prev = node;
+        head = node;
+        size++;
+    }
```

Comment:
delete these methods and add one
`private void linkBefore(T value, Node<T> target)`